txnkv/transaction/pessimistic.go

```diff
@@ -150,156 +168,375 @@ func (action actionPessimisticLock) handleSingleBatch(c *twoPhaseCommitter, bo *
+	regionErr, err := resp.GetRegionError()
+	if err != nil {
+		return true, err
+	}
+	if regionErr != nil {
+		// For other region error and the fake region error, backoff because
+		// there's something wrong.
+		// For the real EpochNotMatch error, don't backoff.
+		if regionErr.GetEpochNotMatch() == nil || locate.IsFakeRegionError(regionErr) {
+			err = bo.Backoff(retry.BoRegionMiss, errors.New(regionErr.String()))
+			if err != nil {
+				return true, err
+			}
+		}
+		same, err := batch.relocate(bo, c.store.GetRegionCache())
+		if err != nil {
+			return true, err
+		}
+		if same {
+			return false, nil
+		}
+		err = c.pessimisticLockMutations(bo, action.LockCtx, action.wakeUpMode, batch.mutations)
+		return true, err
+	}
+	if resp.Resp == nil {
+		return true, errors.WithStack(tikverr.ErrBodyMissing)
+	}
+	lockResp := resp.Resp.(*kvrpcpb.PessimisticLockResponse)
+	if len(lockResp.Results) != 0 {
+		// We use old protocol in this mode. The `Result` field should not be used.
+		return true, errors.New("Pessimistic lock response corrupted")
+	}
+	keyErrs := lockResp.GetErrors()
+	if len(keyErrs) == 0 {
+
+		if action.LockCtx.Stats != nil {
+			action.LockCtx.Stats.MergeReqDetails(diagCtx.reqDuration, batch.region.GetID(), diagCtx.sender.GetStoreAddr(), lockResp.ExecDetailsV2)
+		}
+
+		if batch.isPrimary {
+			// After locking the primary key, we should protect the primary lock from expiring
+			// now in case locking the remaining keys take a long time.
+			c.run(c, action.LockCtx)
+		}
+
+		// Handle the case that the TiKV's version is too old and doesn't support `CheckExistence`.
+		// If `CheckExistence` is set, `ReturnValues` is not set and `CheckExistence` is not supported, skip
+		// retrieving value totally (indicated by `skipRetrievingValue`) to avoid panicking.
+		skipRetrievingValue := !action.ReturnValues && action.CheckExistence && len(lockResp.NotFounds) == 0
+
+		if (action.ReturnValues || action.CheckExistence) && !skipRetrievingValue {
+			action.ValuesLock.Lock()
+			for i, mutation := range mutationsPb {
+				var value []byte
+				if action.ReturnValues {
+					value = lockResp.Values[i]
+				}
+				var exists = !lockResp.NotFounds[i]
+				action.Values[string(mutation.Key)] = kv.ReturnedValue{
+					Value:  value,
+					Exists: exists,
+				}
 			}
+			action.ValuesLock.Unlock()
+		}
+		return true, nil
+	}
+	var locks []*txnlock.Lock
+	for _, keyErr := range keyErrs {
+		// Check already exists error
+		if alreadyExist := keyErr.GetAlreadyExist(); alreadyExist != nil {
+			e := &tikverr.ErrKeyExist{AlreadyExist: alreadyExist}
+			return true, c.extractKeyExistsErr(e)
+		}
+		if deadlock := keyErr.Deadlock; deadlock != nil {
+			return true, errors.WithStack(&tikverr.ErrDeadlock{Deadlock: deadlock})
+		}
+
+		// Extract lock from key error
+		lock, err1 := txnlock.ExtractLockFromKeyErr(keyErr)
+		if err1 != nil {
+			return true, err1
+		}
+		locks = append(locks, lock)
+	}
```

Comment:
There seem to be some duplicates with the implementation in `handlePessimisticLockResponseForceMode`, could it be abstracted to a single function?